pkg/runner/eventapi/eventapi.go

```diff
@@ -0,0 +1,169 @@
+
+	rec := EventReceiver{
+		Events:     make(chan JobEvent, 1000),
+		SocketPath: sockPath,
+		URLPath:    "/events/",
+		ident:      ident,
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc(rec.URLPath, rec.handleEvents)
+	srv := http.Server{Handler: mux}
+	rec.server = &srv
+
+	go func() {
+		errChan <- srv.Serve(listener)
+	}()
+	return &rec, nil
+}
+
+// Close ensures that appropriate resources are cleaned up, such as any unix
+// streaming socket that may be in use. Close must be called.
+func (e *EventReceiver) Close() {
+	e.mutex.Lock()
+	e.stopped = true
+	e.mutex.Unlock()
+	logrus.Debug("event API stopped")
+	e.server.Close()
+	close(e.Events)
+}
+
+func (e *EventReceiver) handleEvents(w http.ResponseWriter, r *http.Request) {
+	logger := logrus.WithFields(logrus.Fields{
+		"component": "eventapi",
+		"job":       e.ident,
+	})
+
+	if r.URL.Path != e.URLPath {
+		http.NotFound(w, r)
+		logger.WithFields(logrus.Fields{
+			"code": "404",
+		}).Infof("path not found: %s\n", r.URL.Path)
+		return
+	}
+
+	if r.Method != http.MethodPost {
+		logger.WithFields(logrus.Fields{
+			"code": "405",
+		}).Infof("method %s not allowed", r.Method)
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	ct := r.Header.Get("content-type")
+	if strings.Split(ct, ";")[0] != "application/json" {
+		logger.WithFields(logrus.Fields{
+			"code": "415",
+		}).Info("wrong content type: %s", ct)
+		w.WriteHeader(http.StatusUnsupportedMediaType)
+		w.Write([]byte("The content-type must be \"application/json\""))
+		return
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		logger.WithFields(logrus.Fields{
+			"code": "500",
+		}).Errorf("%s", err.Error())
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	event := JobEvent{}
+	err = json.Unmarshal(body, &event)
+	if err != nil {
+		logger.WithFields(logrus.Fields{
+			"code": "400",
+		}).Infof("could not deserialize body: %s", err.Error())
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("Could not deserialize body as JSON"))
+		return
+	}
+
+	// Guarantee that the Events channel will not be written to if stopped ==
+	// true, because in that case the channel has been closed.
+	e.mutex.RLock()
+	defer e.mutex.RUnlock()
+	if e.stopped {
+		e.mutex.RUnlock()
+		w.WriteHeader(http.StatusGone)
+		logger.WithFields(logrus.Fields{
+			"code": "410",
+		}).Info("stopped and not accepting additional events for this job")
+		return
+	} else {
+		// ansible-runner sends "status events" and "ansible events". The "status
+		// events" signify a change in the state of ansible-runner itself, which
+		// we're not currently interested in.
+		// https://ansible-runner.readthedocs.io/en/latest/external_interface.html#event-structure
+		if event.UUID == "" {
+			logger.Info("dropping event that is not a JobEvent")
+		} else {
+			// timeout if the channel blocks for too long
+			timeout := time.NewTimer(10 * time.Second)
+			select {
+			case e.Events <- event:
+			case <-timeout.C:
```

Comment:
The reason I didn't do that is this, from the docs on `time.After()`: "The underlying Timer is not recovered by the garbage collector until the timer fires. If efficiency is a concern, use NewTimer instead and call Timer.Stop if the timer is no longer needed."

Since there could be a large volume of these events flowing in quickly, it seemed a shame to leave potentially thousands of timers waiting to fire (just the act of firing has substantial expense) and not getting garbage collected.

Maybe it's a premature optimization, but for a little extra verbosity in the code, it seemed like a reasonable hedge. Thoughts?

Reply:
It might be a premature optimazation, but makes sense to me 😄 I say keep it.

Reply:
:+1: